cmd/gpu_plugin/gpu_plugin_resource_manager.go

```diff
@@ -375,6 +415,66 @@ func containerCards(pod *v1.Pod, gpuUsingContainerIndex int) []string {
 	return nil
 }
 
+func containerCardsAndResources(pod *v1.Pod, gpuUsingContainerIndex int) ([]string, resourceMap) {
+	cards := containerCards(pod, gpuUsingContainerIndex)
+	resources := containerResources(pod, gpuUsingContainerIndex)
+
+	// calculate per card resourceMap (only homogeneous resource consumption is allowed)
+	cardCount := int64(len(cards))
+	cardResources := resourceMap{}
+
+	if cardCount > 0 {
+		for resourceName, value := range resources {
+			// calculate per card resources
+			cardResources[resourceName] = value / cardCount
+		}
+	}
+
+	return cards, cardResources
+}
+
+// addResourceAnnotations adds resource annotations for cgroup purposes to the container response.
+func addResourceAnnotations(cresp *pluginapi.ContainerAllocateResponse, resources resourceMap) error {
+	for _, device := range cresp.Devices {
+		if strings.Contains(device.HostPath, "card") {
+			majorMinor, err := devicePathToMajorMinorStr(device.HostPath)
+			if err != nil {
+				return errors.WithStack(err)
+			}
+
+			for resourceName, value := range resources {
+				lastSeparatorIndex := strings.LastIndex(resourceName, "/")
+				if lastSeparatorIndex == -1 || lastSeparatorIndex == len(resourceName)-1 {
+					return errors.WithStack(err)
+				}
+				resourceSuffix := resourceName[lastSeparatorIndex+1:]
+				valueString := strconv.FormatInt(value, 10)
+				key := "gpu." + resourceSuffix
+				oldVal, ok := cresp.Annotations[key]
+				if ok {
+					cresp.Annotations[key] = oldVal + "," + majorMinor + "=" + valueString
+				} else {
+					cresp.Annotations[key] = majorMinor + "=" + valueString
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func devicePathToMajorMinorStr(path string) (string, error) {
+	lastPathSeparatorIndex := strings.LastIndex(path, "/")
+	if lastPathSeparatorIndex == -1 || lastPathSeparatorIndex == len(path)-1 {
+		return "", errBadPath
+	}
```

Comment:
Remove the latter check, so that empty filename will return errBadCardName instead of errBadPath?